validate.go

```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
 
 	logger.Info("validating request")
 
+	if kind := gjson.GetBytes(payload, "request.object.kind"); kind.String() != "Pod" {
+		return kubewarden.RejectRequest(
+			kubewarden.Message("object is not of kind Pod: rejecting request"),
+			kubewarden.Code(421))
+	}
+
 	data := gjson.GetBytes(
 		payload,
-		"request.object.metadata.name")
+		"request.object.spec.securityContext.sysctls")
 
 	if !data.Exists() {
-		logger.Warn("cannot read object name from metadata: accepting request")
+		logger.Warn("pod doesn't specify sysctls: accepting request")
 		return kubewarden.AcceptRequest()
 	}
-	name := data.String()
 
-	logger.DebugWithFields("validating ingress object", func(e onelog.Entry) {
-		namespace := gjson.GetBytes(payload, "request.object.metadata.namespace").String()
+	logger.DebugWithFields("validating pod object", func(e onelog.Entry) {
+		name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+		namespace := gjson.GetBytes(payload,
+			"request.object.metadata.namespace").String()
 		e.String("name", name)
 		e.String("namespace", namespace)
 	})
 
-	if settings.DeniedNames.Contains(name) {
-		logger.InfoWithFields("rejecting ingress object", func(e onelog.Entry) {
+	// build set of safe sysctls:
+	knownSafeSysctlsArray := getSafeSysctls()
+	knownSafeSysctls := mapset.NewThreadUnsafeSet()
+	for i := 0; i < len(knownSafeSysctlsArray); i++ {
+		knownSafeSysctls.Add(knownSafeSysctlsArray[i])
+	}
+
+	// build set of prefixes from patterns of forbidden sysctls:
+	globForbiddenSysctls := mapset.NewThreadUnsafeSet()
+	for _, elem := range settings.ForbiddenSysctls.ToSlice() {
+		if strings.HasSuffix(elem.(string), "*") {
+			globForbiddenSysctls.Add(strings.TrimSuffix(elem.(string), "*"))
+		}
+	}
+
+	data.ForEach(func(key, value gjson.Result) bool {
+		sysctl := gjson.Get(value.String(), "name").String()
+
+		if settings.ForbiddenSysctls.Contains(sysctl) {
+			err = fmt.Errorf("sysctl %s is on the forbidden list", sysctl)
+			return false // stop iterating
+		}
+
+		// if sysctl matches a pattern, it is forbidden:
+		for _, elem := range globForbiddenSysctls.ToSlice() {
+			if strings.HasPrefix(sysctl, elem.(string)) {
+				err = fmt.Errorf("sysctl %s is on the forbidden list", sysctl)
+				return false // stop iterating
+			}
+		}
+
+		// if sysctl is not on the safe list nor an exception, it is forbidden:
+		if !knownSafeSysctls.Contains(sysctl) &&
+			!settings.AllowedUnsafeSysctls.Contains(sysctl) {
+			err = fmt.Errorf("sysctl %s is not on safe list, nor is in the allowedUnsafeSysctls list",
+				sysctl)
+			return false // stop iterating
+		}
+		return true // continue iterating
+	})
```

Comment:
I would change the block of code with this pseudo code:

```
not_allowed = data <intesection> settings.forbidden
if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

to_eval = data - settings.allowed
not_allowed := []
for sysctl := range to_eval {
  if match(settings.not_allowed_glob, sysctl) {
    not_allowed = append(not_allowed, sysctl)
  }
}

if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

accept

```

This is just a matter of personal taste. We have Set objects with all the things that are allowed and not allowed, I just happen to like to perform "set operations" as much as possible.

Reply:
This does sound clearer, but I don't think it leads to a faster implementation though; right now we are iterating through the pod sysctls and if we find one sysctl that is not good, we stop and reject. With the sets, we would first perform set operations (which, given that the are implemented with maps is going to need the same iterations), and then iterate nevertheless, as we need to check against each blob prefix.